onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -171,6 +171,90 @@ activation.)DOC")
       })
       .SetDoc(R"DOC(Tokenizer divides each string in X into a vector of strings along the last axis. All input strings including attributes are UTF-8 encoded.)DOC");
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Ngram)
```

Comment:
>ONNX_CONTRIB_OPERATOR_SCHEMA [](start = 2, length = 28)

Needs TypeAndShapeInferenceFunction to be specified. Probably fairly simple and could be a lambda.

Reply:
The TypeAndShapeInferenceFunction needs to do the calculation if the output type/shape isn't specified in the model. This needs to happen so the downstream operators get this information. The shape inferencing happens prior to execution.

I believe the latest guidance is that all contrib ops need to provide type/shape inferencing.

---
In reply to: [243097467](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243097467) [](ancestors = 243097467,242792823)

Reply:
I will provide the function and the set the type but the shape is not known until the attributes are processed.

---
In reply to: [243104391](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243104391) [](ancestors = 243104391,243097467,242792823)

Reply:
Attributes are available during type/shape inferencing. See InferenceContextImpl in graph.cc.

---
In reply to: [243109790](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243109790) [](ancestors = 243109790,243104391,243097467,242792823)

Reply:
Will do. For my further education, in Tokenizer the number of tokens is not known until Compute() runs, what would be the right thing to do then?

---
In reply to: [243145505](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243145505) [](ancestors = 243145505,243109790,243104391,243097467,242792823)

Reply:
Some things are impossible. :) Same goes with Reshape - the shape isn't known until runtime so we can't infer the output shape. 

Was only asking for you to update Ngram though. 

---
In reply to: [243375762](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243375762) [](ancestors = 243375762,243145505,243109790,243104391,243097467,242792823)